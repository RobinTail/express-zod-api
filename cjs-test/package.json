{
  "name": "cjs-test",
  "private": true,
  "version": "0.0.0",
  "scripts": {
    "test": "vitest run --globals"
  },
  "dependencies": {
    "express-zod-api": "workspace:*",
    "zod": "catalog:peer"
  }
}
