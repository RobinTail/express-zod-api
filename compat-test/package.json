{
  "name": "compat-test",
  "private": true,
  "type": "module",
  "version": "0.0.0",
  "dependencies": {
    "eslint": "9.0.0",
    "typescript-eslint": "8.0.0",
    "typescript": "~5.7.3"
  },
  "scripts": {
    "pretest": "echo 'client.provide(\"get\", \"/v1/test\", {id: 10});' > sample.ts",
    "test": "eslint --fix && vitest run --globals && rm sample.ts"
  }
}
