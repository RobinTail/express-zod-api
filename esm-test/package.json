{
  "name": "esm-test",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "test": "vitest run --globals"
  }
}
