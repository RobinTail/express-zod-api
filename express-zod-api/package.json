{
  "name": "express-zod-api",
  "version": "23.4.0",
  "description": "A Typescript framework to help you get an API server up and running with I/O schema validation and custom middlewares in minutes.",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/RobinTail/express-zod-api.git"
  },
  "homepage": "https://ez.robintail.cz",
  "author": {
    "name": "Anna Bocharova",
    "url": "https://robintail.cz",
    "email": "me@robintail.cz"
  },
  "bugs": "https://github.com/RobinTail/express-zod-api/issues",
  "funding": "https://github.com/sponsors/RobinTail",
  "scripts": {
    "build": "tsup",
    "postbuild": "yarn pack --filename release.tgz && attw release.tgz && rm release.tgz",
    "pretest": "tsc --noEmit",
    "test": "vitest run --coverage",
    "bench": "vitest bench --run ./bench",
    "prepublishOnly": "eslint && yarn test && yarn build",
    "prepack": "cp ../*.md ../LICENSE ./",
    "postversion": "git push && git push --tags"
  },
  "type": "module",
  "sideEffects": true,
  "main": "dist/index.cjs",
  "types": "dist/index.d.ts",
  "module": "dist/index.js",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.js"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      }
    },
    "./migration": {
      "import": {
        "types": "./migration/index.d.ts",
        "default": "./migration/index.js"
      },
      "require": {
        "types": "./migration/index.d.cts",
        "default": "./migration/index.cjs"
      }
    }
  },
  "files": [
    "dist",
    "migration",
    "*.md"
  ],
  "engines": {
    "node": "^20.9.0 || ^22.0.0 || ^24.0.0"
  },
  "dependencies": {
    "ansis": "^4.0.0",
    "node-mocks-http": "^1.17.2",
    "openapi3-ts": "^4.4.0",
    "ramda": "^0.30.1"
  },
  "peerDependencies": {
    "@types/compression": "^1.7.5",
    "@types/express": "^5.0.0",
    "@types/express-fileupload": "^1.5.0",
    "@types/http-errors": "^2.0.2",
    "compression": "^1.8.0",
    "express": "^5.1.0",
    "express-fileupload": "^1.5.0",
    "http-errors": "^2.0.0",
    "typescript": "^5.1.3",
    "zod": "^3.23.0"
  },
  "peerDependenciesMeta": {
    "@types/compression": {
      "optional": true
    },
    "@types/express": {
      "optional": true
    },
    "@types/express-fileupload": {
      "optional": true
    },
    "@types/http-errors": {
      "optional": true
    },
    "compression": {
      "optional": true
    },
    "express-fileupload": {
      "optional": true
    }
  },
  "devDependencies": {
    "@arethetypeswrong/cli": "^0.18.1",
    "@types/cors": "^2.8.18",
    "@types/depd": "^1.1.36",
    "@types/node-forge": "^1.3.11",
    "@types/ramda": "^0.30.0",
    "@types/semver": "^7.7.0",
    "camelize-ts": "^3.0.0",
    "cors": "^2.8.5",
    "depd": "^2.0.0",
    "node-forge": "^1.3.1",
    "semver": "^7.6.3",
    "snakify-ts": "^2.3.0",
    "tsup": "^8.4.0",
    "undici": "^6.19.8"
  },
  "keywords": [
    "nodejs",
    "api",
    "http",
    "middleware",
    "documentation",
    "json",
    "express",
    "typescript",
    "schema",
    "server",
    "handler",
    "swagger",
    "documentation-tool",
    "openapi",
    "schema-validation",
    "endpoint",
    "openapi-specification",
    "swagger-documentation",
    "zod",
    "validation"
  ]
}
