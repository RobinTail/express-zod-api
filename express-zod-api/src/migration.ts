import {
  ESLintUtils,
  // AST_NODE_TYPES as NT,
  type TSESLint,
  // type TSESTree,
} from "@typescript-eslint/utils"; // eslint-disable-line allowed/dependencies -- special case

// eslint-disable-next-line @typescript-eslint/no-empty-object-type -- @todo
interface Queries {}

type Listener = keyof Queries;

const queries: Record<Listener, string> = {};

const listen = <
  S extends { [K in Listener]: TSESLint.RuleFunction<Queries[K]> },
>(
  subject: S,
) =>
  (Object.keys(subject) as Listener[]).reduce<{ [K: string]: S[Listener] }>(
    (agg, key) =>
      Object.assign(agg, {
        [queries[key]]: subject[key],
      }),
    {},
  );

const v24 = ESLintUtils.RuleCreator.withoutDocs({
  meta: {
    type: "problem",
    fixable: "code",
    schema: [],
    messages: {
      change: "change {{ subject }} from {{ from }} to {{ to }}",
      add: `add {{ subject }} to {{ to }}`,
      move: "move {{ subject }} to {{ to }}",
    },
  },
  defaultOptions: [],
  create: () => listen({}),
});

/**
 * @desc ESLint plugin for migrating to this version (from previous), requires eslint v9 and typescript-eslint v8
 * @deprecated Single-use tool that can be removed and changed regardless SemVer. Remember to delete it after use.
 * @example
 *          // eslint.config.mjs:
 *          import parser from "@typescript-eslint/parser";
 *          import migration from "express-zod-api/migration";
 *          export default [
 *            { languageOptions: {parser}, plugins: {migration} },
 *            { files: ["**\/*.ts"], rules: { "migration/v21": "error" } }
 *          ];
 * */
export default {
  rules: { v24 },
} satisfies TSESLint.Linter.Plugin;
