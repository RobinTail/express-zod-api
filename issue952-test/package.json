{
  "name": "issue952-test",
  "private": true,
  "version": "0.0.0",
  "scripts": {
    "test": "tsc -p tsconfig.json",
    "posttest": "rm *.d.ts"
  },
  "dependencies": {
    "@types/express": "catalog:peer",
    "@types/express-fileupload": "catalog:peer",
    "express-zod-api": "workspace:*",
    "typescript": "catalog:peer",
    "zod": "catalog:peer"
  }
}
