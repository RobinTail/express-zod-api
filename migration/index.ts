import {
  ESLintUtils,
  AST_NODE_TYPES as NT,
  type TSESLint,
  type TSESTree,
} from "@typescript-eslint/utils"; // eslint-disable-line allowed/dependencies -- assumed transitive dependency

type NamedProp = TSESTree.PropertyNonComputedName & {
  key: TSESTree.Identifier;
};

interface Queries {
  zod: TSESTree.ImportDeclaration;
  dateInOutExample: NamedProp;
  getExamples: TSESTree.CallExpression;
}

type Listener = keyof Queries;

const queries: Record<Listener, string> = {
  zod: `${NT.ImportDeclaration}[source.value='zod/v4']`,
  dateInOutExample:
    `${NT.CallExpression}[callee.object.name='ez'][callee.property.name=/date(In|Out)/] >` +
    `${NT.ObjectExpression} > ${NT.Property}[key.name='example']`,
  getExamples: `${NT.CallExpression}[callee.name='getExamples']`,
};

const listen = <
  S extends { [K in Listener]: TSESLint.RuleFunction<Queries[K]> },
>(
  subject: S,
) =>
  (Object.keys(subject) as Listener[]).reduce<{ [K: string]: S[Listener] }>(
    (agg, key) =>
      Object.assign(agg, {
        [queries[key]]: subject[key],
      }),
    {},
  );

// eslint-disable-next-line no-restricted-syntax -- substituted by TSDOWN and vitest
const ruleName = `v${process.env.TSDOWN_VERSION?.split(".")[0] ?? "0"}`; // fail-safe for bumpp

const theRule = ESLintUtils.RuleCreator.withoutDocs({
  meta: {
    type: "problem",
    fixable: "code",
    schema: [],
    messages: {
      change: "change {{ subject }} from {{ from }} to {{ to }}",
      add: "add {{ subject }} to {{ to }}",
      move: "move {{ subject }} to {{ to }}",
      remove: "remove {{ subject }}",
    },
  },
  defaultOptions: [],
  create: (ctx) =>
    listen({
      zod: (node) =>
        ctx.report({
          node: node.source,
          messageId: "change",
          data: { subject: "import", from: "zod/v4", to: "zod" },
          fix: (fixer) => fixer.replaceText(node.source, `"zod"`),
        }),
      dateInOutExample: (node) =>
        ctx.report({
          node,
          messageId: "change",
          data: { subject: "property", from: "example", to: "examples" },
          fix: (fixer) =>
            fixer.replaceText(
              node,
              `examples: [${ctx.sourceCode.getText(node.value)}]`,
            ),
        }),
      getExamples: (node) =>
        ctx.report({
          node,
          messageId: "change",
          data: {
            subject: "method",
            from: "getExamples()",
            to: ".meta()?.examples || []",
          },
          fix: (fixer) =>
            fixer.replaceText(
              node,
              `(${ctx.sourceCode.getText(node.arguments[0])}.meta()?.examples || [])`,
            ),
        }),
    }),
});

export default {
  rules: { [ruleName]: theRule } as Record<`v${number}`, typeof theRule>,
} satisfies TSESLint.Linter.Plugin;
