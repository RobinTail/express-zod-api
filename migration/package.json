{
  "name": "@express-zod-api/migration",
  "version": "25.0.0-beta.2",
  "license": "MIT",
  "description": "Migration scripts for express-zod-api",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/RobinTail/express-zod-api.git"
  },
  "author": {
    "name": "Anna Bocharova",
    "url": "https://robintail.cz",
    "email": "me@robintail.cz"
  },
  "bugs": "https://github.com/RobinTail/express-zod-api/issues",
  "funding": "https://github.com/sponsors/RobinTail",
  "scripts": {
    "pretest": "tsc --noEmit",
    "test": "vitest run --globals",
    "build": "tsup",
    "postbuild": "attw --pack --profile esm-only",
    "prepublishOnly": "eslint && pnpm test && pnpm build"
  },
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "module": "dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
  "files": [
    "dist",
    "*.md"
  ],
  "peerDependencies": {
    "eslint": "^9.0.0",
    "typescript-eslint": "^8.0.0"
  },
  "devDependencies": {
    "@typescript-eslint/rule-tester": "catalog:dev"
  }
}
