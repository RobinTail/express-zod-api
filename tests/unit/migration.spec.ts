import { RuleTester } from "@typescript-eslint/rule-tester";
import migration from "../../src/migration";
import parser from "@typescript-eslint/parser";
import { version } from "../../package.json";

RuleTester.afterAll = afterAll;
RuleTester.describe = describe;
RuleTester.it = it;

const tester = new RuleTester({
  languageOptions: { parser },
});

describe("Migration", () => {
  test("should consist of one rule being the major version of the package", () => {
    expect(migration.rules).toHaveProperty(`v${version.split(".")[0]}`);
    expect(migration).toMatchSnapshot();
  });
});

tester.run("v21", migration.rules.v21, {
  valid: [{ code: `import {} from "express-zod-api"` }],
  invalid: [],
});
