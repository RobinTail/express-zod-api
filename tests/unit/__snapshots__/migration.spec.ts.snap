// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Migration > should consist of one rule being the major version of the package 1`] = `
{
  "rules": {
    "v21": {
      "create": [Function],
      "defaultOptions": [],
      "meta": {
        "fixable": "code",
        "messages": {},
        "schema": [],
        "type": "problem",
      },
    },
  },
}
`;
